aten/src/ATen/core/Generator.cpp

```diff
@@ -0,0 +1,316 @@
+#include "ATen/core/Generator.h"
+
+namespace at {
+
+/*
+* GeneratorState copy and assign constructor implementations
+*/
+GeneratorState::GeneratorState(const GeneratorState& other) {
+  this->device = other.device;
+  this->current_seed = other.current_seed;
+  this->device_type = other.device_type;
+  this->philox_offset_per_thread.exchange(other.philox_offset_per_thread);
+  new (&(this->cpu_engine)) std::mt19937_64(other.cpu_engine);
+}
+
+GeneratorState& GeneratorState::operator=(const GeneratorState& other) {
+  if(&other == this) return *this;
+  this->device = other.device;
+  this->current_seed = other.current_seed;
+  this->device_type = other.device_type;
+  this->philox_offset_per_thread.exchange(other.philox_offset_per_thread);
+  new (&(this->cpu_engine)) std::mt19937_64(other.cpu_engine);
```

Comment:
See above comment, but I also want to note that this placement-new isn't even right: `cpu_engine` previously contained a valid mt19937_64 engine and you have to placement-delete it first before you reuse the memory for something else.

Reply:
Fixed this by using mt19937 copy constructor.